### PR TITLE
Fix #117: Implement faceted search.

### LINKED ISSUE #117: Implement faceted search
With kinto-elasticsearch, we could perform ElasticSearch queries to retrieve refined available filters from a build search:

```json
{
  "size": 0,
  "query": {
    "bool": {
      "must" : [
        {"match": {"source.product": "fennec"}},
        {"match": {"target.channel": "nightly"}},
        {"match": {"target.locale": "te"}}
      ]
    }
  },
  "aggregations" : {
    "product_filters" : {
      "terms": {"field" : "source.product"}
    },
    "channel_filters" : {
      "terms": {"field" : "target.channel"}
    },
    "platform_filters" : {
      "terms": {"field" : "target.platform"}
    },
    "version_filters" : {
      "terms": {"field" : "target.version"}
    },
    "locale_filters" : {
      "terms": {"field" : "target.locale"}
    }
  }
}
```

Example: to see what facets remain when searching for Fennec nightly builds for the `te` language, POSTing this query to `https://kinto-ota.dev.mozaws.net/v1/buckets/build-hub/collections/releases/search` will give the following output:

```json
{
  "timed_out": false,
  "hits": {
    "hits": [],
    "total": 1,
    "max_score": 0.0
  },
  "aggregations": {
    "version_filters": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "55.0a1"
        }
      ],
      "sum_other_doc_count": 0,
      "doc_count_error_upper_bound": 0
    },
    "channel_filters": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "nightly"
        }
      ],
      "sum_other_doc_count": 0,
      "doc_count_error_upper_bound": 0
    },
    "product_filters": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "fennec"
        }
      ],
      "sum_other_doc_count": 0,
      "doc_count_error_upper_bound": 0
    },
    "locale_filters": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "te"
        }
      ],
      "sum_other_doc_count": 0,
      "doc_count_error_upper_bound": 0
    },
    "platform_filters": {
      "buckets": [
        {
          "doc_count": 1,
          "key": "android"
        },
        {
          "doc_count": 1,
          "key": "arm"
        }
      ],
      "sum_other_doc_count": 0,
      "doc_count_error_upper_bound": 0
    }
  },
  "took": 5,
  "_shards": {
    "total": 5,
    "failed": 0,
    "successful": 5
  }
}
```

Which is all we need. Now we'll have to forge custom HTTP requests as elm-kinto doesn't support the `search` endpoint exposed by kinto-elasticsearch.

### PR DESCRIPTION
**Pull request by n1k0**
Refs #117, WiP.

### Todo

- [x] create encoder and decoders for ES query and results
- [x] expose some HTTP getter for results
- [x] remove old filters form
- [x] handle product, version, platform, channel and locale filters
- [x] handle buildId search
- [x] list results from ES, not kinto
- [x] back to live filtering, no form submit
- [x] cleanup all kinto related code
- [x] sort versions appropriately
- [x] fix routing, make `page` an url parameter

### CODE REVIEW

**Comment by n1k0** [comment]
Note: Aggregates are not accurate because of https://github.com/mozilla-services/buildhub/issues/125.

File: ui/src/ElasticSearch.elm [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        "channel" ->
+            Just "target.channel"
+
+        "locale" ->
+            Just "target.locale"
+
+        "version" ->
+            Just "target.version"
+
+        "platform" ->
```
Comment (glasserc) [inline ui/src/ElasticSearch.elm]: Hmm, this function has a code smell to me. My first thought is that filters should be an enum. But then I looked at the context where this is used and I wonder if maybe this is even necessary? Why don't we just use the JSON names at the call site?

File: ui/src/ElasticSearch.elm [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        , ( "version", version )
+        , ( "platform", platform )
+        ]
+            |> List.foldr refineFilters []
+            |> List.map encodeFilter
+            |> Encode.list
+
+
+encodeQuery : Filters -> Int -> Int -> Encode.Value
+encodeQuery filters size page =
```
Comment (glasserc) [inline ui/src/ElasticSearch.elm]: How about renaming the variable `size` to `pageSize` or `resultsPerPage` or `limit` or something?

File: ui/src/View.elm [hunk 3]
```diff
@@ -24,20 +23,23 @@ view model =
...
+            , case facets of
+                Just facets ->
+                    div []
+                        [ paginationView facets settings.pageSize page
+                        , div [] <| List.map recordView facets.hits
+                        , paginationView facets settings.pageSize page
+                        ]
+
+                Nothing ->
+                    spinner
```
Comment (glasserc) [inline ui/src/View.elm]: :+1: for moving to `Maybe` here. Do we still need `loading`?

File: ui/src/View.elm [hunk 4]
```diff
@@ -24,20 +23,23 @@ view model =
...
+            ]
         , div [ class "col-sm-3" ]
-            [ filtersView filters filterValues
+            [ filtersView facets filters filterValues
```
Comment (glasserc) [inline ui/src/View.elm]: I'm not crazy about `filtersView` having to take a `Maybe` for facets. Maybe do another `case` here to put a spinner here too?

File: ui/src/View.elm [hunk 5]
```diff
@@ -130,31 +132,53 @@ errorView err =
...
+                    [ div [ class "btn-group" ]
+                        [ button
+                            [ class "btn btn-default"
+                            , onClick LoadPreviousPage
+                            , disabled (page == 1)
+                            ]
+                            [ text "Previous" ]
+                        , button
+                            [ class "btn btn-default"
+                            , onClick LoadNextPage
```
Comment (glasserc) [inline ui/src/View.elm]: Any condition on this one? (Disabled if page == Math.ceil (total/pageSize)?)

File: ui/src/ElasticSearch.elm (hunk 2 above)
Comment (n1k0) [inline ui/src/ElasticSearch.elm]: Good idea

File: ui/src/View.elm (hunk 3 above)
Comment (n1k0) [inline ui/src/View.elm]: :+1:

File: ui/src/View.elm (hunk 4 above)
Comment (n1k0) [inline ui/src/View.elm]: :+1:

File: ui/src/ElasticSearch.elm (hunk 1 above)
Comment (n1k0) [inline ui/src/ElasticSearch.elm]: Yeah this was hacked in a hurry. I'd love filters to be an array of json field names indeed, but maybe I can address that in a next PR.

File: ui/src/View.elm (hunk 5 above)
Comment (n1k0) [inline ui/src/View.elm]: Updated.

**Pull request merged**